storage/src/main/java/tech/pegasys/artemis/storage/V2MapDatabase.java

```diff
@@ -0,0 +1,291 @@
+    } catch (IOException e) {
+      STDOUT.log(Level.WARN, "Failed to clear old database");
+    }
+    return new V2MapDatabase(DBMaker.fileDB(databaseFile));
+  }
+
+  public static Database createInMemory() {
+    return new V2MapDatabase(DBMaker.memoryDB());
+  }
+
+  private V2MapDatabase(final Maker dbMaker) {
+    db = dbMaker.transactionEnable().make();
+    time = db.atomicVar("time", new UnsignedLongSerializer()).createOrOpen();
+    genesisTime = db.atomicVar("genesisTime", new UnsignedLongSerializer()).createOrOpen();
+    justifiedCheckpoint =
+        db.atomicVar("justifiedCheckpoint", new MapDBSerializer<>(Checkpoint.class)).createOrOpen();
+    bestJustifiedCheckpoint =
+        db.atomicVar("bestJustifiedCheckpoint", new MapDBSerializer<>(Checkpoint.class))
+            .createOrOpen();
+    finalizedCheckpoint =
+        db.atomicVar("finalizedCheckpoint", new MapDBSerializer<>(Checkpoint.class)).createOrOpen();
+
+    finalizedRootsBySlot =
+        db.treeMap("finalizedRootsBySlot", new UnsignedLongSerializer(), new Bytes32Serializer())
+            .createOrOpen();
+    finalizedBlocksByRoot =
+        db.hashMap(
+                "finalizedBlocksByRoot",
+                new Bytes32Serializer(),
+                new MapDBSerializer<>(BeaconBlock.class))
+            .createOrOpen();
+    finalizedStatesByRoot =
+        db.hashMap(
+                "finalizedStatsByRoot",
+                new Bytes32Serializer(),
+                new MapDBSerializer<>(BeaconState.class))
+            .createOrOpen();
+
+    hotBlocksByRoot =
+        db.hashMap(
+                "hotBlocksByRoot",
+                new Bytes32Serializer(),
+                new MapDBSerializer<>(BeaconBlock.class))
+            .createOrOpen();
+    hotStatesByRoot =
+        db.hashMap(
+                "hotStatesByRoot",
+                new Bytes32Serializer(),
+                new MapDBSerializer<>(BeaconState.class))
+            .createOrOpen();
+
+    checkpointStates =
+        db.hashMap(
+                "checkpointStates",
+                new MapDBSerializer<>(Checkpoint.class),
+                new MapDBSerializer<>(BeaconState.class))
+            .createOrOpen();
+
+    latestMessages =
+        db.hashMap(
+                "latestMessages",
+                new UnsignedLongSerializer(),
+                new MapDBSerializer<>(Checkpoint.class))
+            .createOrOpen();
+
+    // Recreate hotRootsBySlotCache
+    hotBlocksByRoot.forEach(this::addToHotRootsBySlotCache);
+  }
+
+  @Override
+  public void storeGenesis(final Store store) {
+    time.set(store.getTime());
+    genesisTime.set(store.getGenesisTime());
+    justifiedCheckpoint.set(store.getJustifiedCheckpoint());
+    finalizedCheckpoint.set(store.getFinalizedCheckpoint());
+    bestJustifiedCheckpoint.set(store.getBestJustifiedCheckpoint());
+    store
+        .getBlockRoots()
+        .forEach(
+            root -> {
+              addHotBlock(root, store.getBlock(root));
```

Comment:
The only block at genesis is the genesis block and it is finalized by definition, so adding it to hot blocks might be a wrong move.

Reply:
same for the genesis block's state

Reply:
Good spot. It turns out we need to put them in both.  The genesis block itself is finalised but it's also required in memory or if we restarted immediately after genesis the `Store` we rebuilt to work off of would actually be empty.